### PR TITLE
Drop unused variables in observation dataset to reduce memory usage of corresponding dataframe

### PR DESCRIPTION
**Pull request by dwfncar**
I have tested this within the unit tests in the develop_testsuite branch with test_obs.py,
which demonstrates reduced memory usage after observertion.obs_to_df.
The YAML parameter drop_unused_vars can be added to an observation dataset.
Need to test on larger examples.

```yaml
obs:
    test_obs:
        filename: test_obs.nc
        obs_type: pt_sfc
        drop_unused_vars: True
        variables:
            A_obs:
                units: 'Units of A'
                unit_scale: 1
                unit_scale_method: '*'
            B_obs:
                units: 'Units of B' 
                unit_scale: 1
                unit_scale_method: '*'
```

### CODE REVIEW

File: melodies_monet/driver.py [hunk 1]
```diff
@@ -175,8 +176,20 @@ def open_obs(self):
...
 
+        if self.drop_unused_vars:
+            self.drop_unused()
+
         self.mask_and_scale()  # mask and scale values from the control values
 
+    def drop_unused(self):
+
+        # drop unused variables
+        print()
```
Comment (zmoon) [inline melodies_monet/driver.py:187]:
```suggestion
```

File: melodies_monet/driver.py [hunk 2]
```diff
@@ -537,6 +555,9 @@ def open_obs(self):
                 o.file = self.control_dict['obs'][obs]['filename']
                 if 'variables' in self.control_dict['obs'][obs].keys():
                     o.variable_dict = self.control_dict['obs'][obs]['variables']
+                if 'drop_unused_vars' in self.control_dict['obs'][obs].keys():
+                    o.drop_unused_vars \
+                        = self.control_dict['obs'][obs]['drop_unused_vars']
```
Comment (zmoon) [inline melodies_monet/driver.py:560]:
```suggestion
                    o.drop_unused_vars = self.control_dict['obs'][obs]['drop_unused_vars']
```
Many lines in MM are longer than this, don't think need to break

File: melodies_monet/driver.py [hunk 3]
```diff
@@ -219,7 +232,12 @@ def obs_to_df(self):
         -------
         None
         """
-        self.obj = self.obj.to_dataframe().reset_index().drop(['x', 'y'], axis=1)
+        try:
+            self.obj = self.obj.to_dataframe().reset_index().drop(['x', 'y'], axis=1)
+        except:
+            self.obj = self.obj.to_dataframe().reset_index()
+        print(self.obj.info())
+        print(self.obj.memory_usage())
```
Comment (zmoon) [inline melodies_monet/driver.py:240]: will these be removed?

**Review by zmoon — COMMENTED** [review]
@dwfncar some questions:
* Can we get the test you mentioned in this PR?
* Model dataset(s) as well? Seems like that could be done the same way.

**Comment by zmoon** [comment]
@dwfncar I edited your original comment to block and highlight the YAML

**Comment by dwfncar** [comment]
Have some more work to do on this:
Running our standard WRF-Chem AirNow example,
I see that there are required variables not explicitly listed in the YAML that are being dropped.

observation.drop_unused:dropping:site
observation.drop_unused:dropping:utcoffset
observation.drop_unused:dropping:units
observation.drop_unused:dropping:cmsa_name
observation.drop_unused:dropping:msa_code
observation.drop_unused:dropping:msa_name
observation.drop_unused:dropping:state_name
observation.drop_unused:dropping:epa_region
observation.drop_unused:dropping:time_local
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7916526 entries, 0 to 7916525
Data columns (total 6 columns):
 #   Column     Dtype         
---  ------     -----         
 0   time       datetime64[ns]
 1   OZONE      float64       
 2   PM2.5      float64       
 3   latitude   float64       
 4   longitude  float64       
 5   siteid     object        
dtypes: datetime64[ns](1), float64(4), object(1)
memory usage: 362.4+ MB
None
Index             128
time         63332208
OZONE        63332208
PM2.5        63332208
latitude     63332208
longitude    63332208
siteid       63332208
dtype: int64
[########################################] | 100% Completed |  0.1s
[########################################] | 100% Completed |  0.1s
[########################################] | 100% Completed |  0.2s
[########################################] | 100% Completed |  0.3s
Traceback (most recent call last):
  File "/Users/fillmore/EarthSystem/Examples/run_mm_overview.py", line 17, in <module>
    an.plotting()
  File "/Users/fillmore/EarthSystem/MELODIES-MONET/melodies_monet/driver.py", line 772, in plotting
    pairdf = pairdf.set_index(grp_dict['data_proc']['ts_select_time'])
  File "/opt/miniconda3/envs/melodies-monet/lib/python3.9/site-packages/pandas/util/_decorators.py", line 311, in wrapper
    return func(*args, **kwargs)
  File "/opt/miniconda3/envs/melodies-monet/lib/python3.9/site-packages/pandas/core/frame.py", line 5451, in set_index
    raise KeyError(f"None of {missing} are in the columns")
KeyError: "None of ['time_local'] are in the columns"

**Comment by zmoon** [comment]
Just noting that yesterday we discussed just focusing on AERONET and AirNow for now for this, with a focus on AirNow since that is the one that doesn't say "under development" in the MM docs.

**Comment by zmoon** [comment]
How about we have a standard list of site metadata ones to save (like this PR already has), but then also allow saving extra ones from control file, like `vars_nodrop: [a, b, c]` @dwfncar

**Comment by zmoon** [comment]
<details><summary>All variables in sample MM AERONET file</summary>

```
['340-440_angstrom_exponent',
 '380-500_angstrom_exponent',
 '440-675_angstrom_exponent',
 '440-675_angstrom_exponent[polar]',
 '440-870_angstrom_exponent',
 '500-870_angstrom_exponent',
 'aeronet_instrument_number',
 'aod_1020nm',
 'aod_11100nm',
 'aod_1630nm',
 'aod_1640nm',
 'aod_380nm',
 'aod_400nm',
 'aod_412nm',
 'aod_443nm',
 'aod_490nm',
 'aod_500nm',
 'aod_510nm',
 'aod_532nm',
 'aod_550nm',
 'aod_551nm',
 'aod_555nm',
 'aod_560nm',
 'aod_620nm',
 'aod_660nm',
 'aod_667nm',
 'aod_675nm',
 'aod_681nm',
 'aod_709nm',
 'aod_779nm',
 'aod_860nm',
 'aod_865nm',
 'aod_870nm',
 'day_of_year',
 'day_of_year(fraction)',
 'elevation',
 'exact_wavelengths_of_aod(um)_1020nm',
 'exact_wavelengths_of_aod(um)_1640nm',
 'exact_wavelengths_of_aod(um)_340nm',
 'exact_wavelengths_of_aod(um)_380nm',
 'exact_wavelengths_of_aod(um)_400nm',
 'exact_wavelengths_of_aod(um)_412nm',
 'exact_wavelengths_of_aod(um)_440nm',
 'exact_wavelengths_of_aod(um)_443nm',
 'exact_wavelengths_of_aod(um)_490nm',
 'exact_wavelengths_of_aod(um)_500nm',
 'exact_wavelengths_of_aod(um)_510nm',
 'exact_wavelengths_of_aod(um)_532nm',
 'exact_wavelengths_of_aod(um)_551nm',
 'exact_wavelengths_of_aod(um)_555nm',
 'exact_wavelengths_of_aod(um)_560nm',
 'exact_wavelengths_of_aod(um)_620nm',
 'exact_wavelengths_of_aod(um)_667nm',
 'exact_wavelengths_of_aod(um)_675nm',
 'exact_wavelengths_of_aod(um)_681nm',
 'exact_wavelengths_of_aod(um)_709nm',
 'exact_wavelengths_of_aod(um)_779nm',
 'exact_wavelengths_of_aod(um)_865nm',
 'exact_wavelengths_of_aod(um)_870nm',
 'exact_wavelengths_of_pw(um)_935nm',
 'latitude',
 'longitude',
 'no2(dobson)',
 'number_of_wavelengths',
 'optical_air_mass',
 'ozone(dobson)',
 'precipitable_water(cm)',
 'sensor_temperature(degrees_c)',
 'siteid',
 'solar_zenith_angle(degrees)',
 'time',
 'triplet_variability_1020',
 'triplet_variability_1640',
 'triplet_variability_340',
 'triplet_variability_380',
 'triplet_variability_400',
 'triplet_variability_412',
 'triplet_variability_440',
 'triplet_variability_443',
 'triplet_variability_490',
 'triplet_variability_500',
 'triplet_variability_510',
 'triplet_variability_532',
 'triplet_variability_551',
 'triplet_variability_555',
 'triplet_variability_560',
 'triplet_variability_620',
 'triplet_variability_667',
 'triplet_variability_675',
 'triplet_variability_681',
 'triplet_variability_709',
 'triplet_variability_779',
 'triplet_variability_865',
 'triplet_variability_870',
 'triplet_variability_precipitable_water(cm)',
 'x']
```
</details>

<details><summary>All variables in sample MM AirNow file</summary>

```
['BARPR',
 'BC',
 'CO',
 'NO',
 'NO2',
 'NO2Y',
 'NOX',
 'NOY',
 'OZONE',
 'PM10',
 'PM2.5',
 'PRECIP',
 'RHUM',
 'RWD',
 'RWS',
 'SO2',
 'SRAD',
 'TEMP',
 'UV-AETH',
 'WD',
 'WS',
 'cmsa_name',
 'epa_region',
 'latitude',
 'longitude',
 'msa_code',
 'msa_name',
 'site',
 'siteid',
 'state_name',
 'time',
 'time_local',
 'units',
 'utcoffset',
 'x']
```
</details>

**Comment by dwfncar** [comment]
Linking issue to this PR: resolves #112

**Comment by dwfncar** [comment]
This branch is well out of date with latest driver.  Given that the new method drop_unused is short and needs to reworked in any event, I am closing this PR.  Will re-branch from develop once we decide where and how dropping unused variables should occur ...

**Pull request closed without merging**